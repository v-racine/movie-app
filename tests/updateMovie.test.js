const { Config } = require('../src/config/config');

Config.Get(process.env);

const AppFactory = require('../src/app');
const request = require('supertest');

describe('update a movie', () => {
  const mockPgClient = {};

  const app = AppFactory({
    pgClient: mockPgClient,
  });

  describe('when: a user wants to view the form to edit a movie', () => {
    let rsp;
    let id = 21;

    beforeEach(async () => {
      rsp = await request(app).get(`/movies/update/${id}`).send();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: we return the view of the form', async () => {
      const text = rsp.text;
      expect(text).toMatchSnapshot();

      const status = rsp.status;
      expect(status).toBe(200);
    });
  });

  describe('when: a user does NOT enter a valid title', () => {
    let rsp;

    let movie = {
      title: '',
      year: 1991,
      runTime: 122,
    };

    beforeEach(async () => {
      mockPgClient.dbQuery = jest.fn();

      rsp = await request(app)
        .post(`/movies/create`)
        .send(`title=${movie.title}&year=${movie.year}&runTime=${movie.runTime}`);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test("then: we return 'Title must be between 1 and 250 characters'", async () => {
      const text = rsp.text;
      expect(text).toMatchSnapshot();

      const status = rsp.status;
      expect(status).toBe(200);
    });
  });

  describe('when: a user does NOT enter a valid year', () => {
    let rsp;

    let movie = {
      title: 'Point Break',
      year: 21,
      runTime: 122,
    };

    beforeEach(async () => {
      mockPgClient.dbQuery = jest.fn();

      rsp = await request(app)
        .post(`/movies/create`)
        .send(`title=${movie.title}&year=${movie.year}&runTime=${movie.runTime}`);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test("then: we return 'Year must be a 4-digit number'", async () => {
      const text = rsp.text;
      expect(text).toMatchSnapshot();

      const status = rsp.status;
      expect(status).toBe(200);
    });
  });

  describe('when: a user does NOT enter a valid run time', () => {
    let rsp;

    let movie = {
      title: 'Point Break',
      year: 1991,
      runTime: '2h2',
    };

    beforeEach(async () => {
      mockPgClient.dbQuery = jest.fn();

      rsp = await request(app)
        .post(`/movies/create`)
        .send(`title=${movie.title}&year=${movie.year}&runTime=${movie.runTime}`);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test("then: we return 'Run time must be in minutes'", async () => {
      const text = rsp.text;
      expect(text).toMatchSnapshot();

      const status = rsp.status;
      expect(status).toBe(200);
    });
  });

  describe('when: a user does NOT enter a valid title NOR a valid year', () => {
    let rsp;

    let movie = {
      title: '',
      year: 'year',
      runTime: 122,
    };

    beforeEach(async () => {
      mockPgClient.dbQuery = jest.fn();

      rsp = await request(app)
        .post(`/movies/create`)
        .send(`title=${movie.title}&year=${movie.year}&runTime=${movie.runTime}`);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test("then: we return both 'Title must be between 1 and 250 characters' AND 'Year must be a 4-digit number'", async () => {
      const text = rsp.text;
      expect(text).toMatchSnapshot();

      const status = rsp.status;
      expect(status).toBe(200);
    });
  });

  describe('when: a user successfully submits the edit-movie form', () => {
    let rsp;
    let id = '21';

    let old = {
      id: id,
      title: 'Playtime',
      year: 1976,
      runTime: 100,
    };

    let updated = {
      id: id,
      title: 'Playtime',
      year: 1967,
      runTime: 115,
    };

    beforeEach(async () => {
      mockPgClient.dbQuery = jest.fn().mockImplementationOnce(() => {
        return new Promise((resolve) => {
          resolve({ rows: [old] });
        });
      });

      rsp = await request(app)
        .post(`/movies/update/${id}`)
        .send(`title=${updated.title}&year=${updated.year}&runTime=${updated.runTime}`);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('then: we update the movie & redirect to the movies page', async () => {
      expect(mockPgClient.dbQuery).toHaveBeenCalledTimes(2);

      expect(mockPgClient.dbQuery).toHaveBeenNthCalledWith(
        1,
        'SELECT * FROM movies WHERE id = $1',
        id,
      );

      expect(mockPgClient.dbQuery).toHaveBeenNthCalledWith(
        2,
        'UPDATE movies SET movie_title = $1, movie_year = $2, run_time = $3 WHERE id = $4',
        updated.title,
        updated.year,
        updated.runTime,
        id,
      );

      const status = rsp.status;
      expect(status).toBe(302);
    });
  });
});
