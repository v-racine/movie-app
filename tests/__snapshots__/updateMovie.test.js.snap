// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update a movie when: a user does NOT enter a valid run time then: we return 'Run time must be in minutes' 1`] = `
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <title>
    </title>
    <meta charset="UTF-8">
    <link rel="stylesheet"
          href="/stylesheets/application.css"
    >
  </head>
  <body>
    <header>
      <h1>
        <a href="/">
          The Ultimate Movie App
        </a>
      </h1>
    </header>
    <main>
      <ul>
        <li class="flash error">
          Run time must be in minutes
        </li>
      </ul>
      <form action="/movies/create"
            method="post"
      >
        <dl>
          <dt>
            <label for="title">
              Enter the title of a new movie:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="title"
                   name="title"
                   placeholder="Movie Title"
                   value="Point Break"
            >
          </dd>
          <dt>
            <label for="year">
              Enter the movie's year of release:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="year"
                   name="year"
                   placeholder="Release Year"
                   value="1991"
            >
          </dd>
          <dt>
            <label for="runTime">
              Enter the movie's run time:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="runTime"
                   name="runTime"
                   placeholder="Run time"
                   value="null"
            >
          </dd>
        </dl>
        <fieldset class="actions">
          <input type="submit"
                 value="Add Movie"
          >
          <input type="reset"
                 value="Clear Form"
          >
        </fieldset>
      </form>
    </main>
  </body>
</html>
`;

exports[`update a movie when: a user does NOT enter a valid title NOR a valid year then: we return both 'Title must be between 1 and 250 characters' AND 'Year must be a 4-digit number' 1`] = `
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <title>
    </title>
    <meta charset="UTF-8">
    <link rel="stylesheet"
          href="/stylesheets/application.css"
    >
  </head>
  <body>
    <header>
      <h1>
        <a href="/">
          The Ultimate Movie App
        </a>
      </h1>
    </header>
    <main>
      <ul>
        <li class="flash error">
          Movie title must be between 1 and 250 characters
        </li>
        <li class="flash error">
          Year must be a 4-digit number
        </li>
      </ul>
      <form action="/movies/create"
            method="post"
      >
        <dl>
          <dt>
            <label for="title">
              Enter the title of a new movie:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="title"
                   name="title"
                   placeholder="Movie Title"
                   value
            >
          </dd>
          <dt>
            <label for="year">
              Enter the movie's year of release:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="year"
                   name="year"
                   placeholder="Release Year"
                   value="null"
            >
          </dd>
          <dt>
            <label for="runTime">
              Enter the movie's run time:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="runTime"
                   name="runTime"
                   placeholder="Run time"
                   value="122"
            >
          </dd>
        </dl>
        <fieldset class="actions">
          <input type="submit"
                 value="Add Movie"
          >
          <input type="reset"
                 value="Clear Form"
          >
        </fieldset>
      </form>
    </main>
  </body>
</html>
`;

exports[`update a movie when: a user does NOT enter a valid title then: we return 'Title must be between 1 and 250 characters' 1`] = `
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <title>
    </title>
    <meta charset="UTF-8">
    <link rel="stylesheet"
          href="/stylesheets/application.css"
    >
  </head>
  <body>
    <header>
      <h1>
        <a href="/">
          The Ultimate Movie App
        </a>
      </h1>
    </header>
    <main>
      <ul>
        <li class="flash error">
          Movie title must be between 1 and 250 characters
        </li>
      </ul>
      <form action="/movies/create"
            method="post"
      >
        <dl>
          <dt>
            <label for="title">
              Enter the title of a new movie:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="title"
                   name="title"
                   placeholder="Movie Title"
                   value
            >
          </dd>
          <dt>
            <label for="year">
              Enter the movie's year of release:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="year"
                   name="year"
                   placeholder="Release Year"
                   value="1991"
            >
          </dd>
          <dt>
            <label for="runTime">
              Enter the movie's run time:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="runTime"
                   name="runTime"
                   placeholder="Run time"
                   value="122"
            >
          </dd>
        </dl>
        <fieldset class="actions">
          <input type="submit"
                 value="Add Movie"
          >
          <input type="reset"
                 value="Clear Form"
          >
        </fieldset>
      </form>
    </main>
  </body>
</html>
`;

exports[`update a movie when: a user does NOT enter a valid year then: we return 'Year must be a 4-digit number' 1`] = `
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <title>
    </title>
    <meta charset="UTF-8">
    <link rel="stylesheet"
          href="/stylesheets/application.css"
    >
  </head>
  <body>
    <header>
      <h1>
        <a href="/">
          The Ultimate Movie App
        </a>
      </h1>
    </header>
    <main>
      <ul>
        <li class="flash error">
          Year must be a 4-digit number
        </li>
      </ul>
      <form action="/movies/create"
            method="post"
      >
        <dl>
          <dt>
            <label for="title">
              Enter the title of a new movie:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="title"
                   name="title"
                   placeholder="Movie Title"
                   value="Point Break"
            >
          </dd>
          <dt>
            <label for="year">
              Enter the movie's year of release:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="year"
                   name="year"
                   placeholder="Release Year"
                   value="21"
            >
          </dd>
          <dt>
            <label for="runTime">
              Enter the movie's run time:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="runTime"
                   name="runTime"
                   placeholder="Run time"
                   value="122"
            >
          </dd>
        </dl>
        <fieldset class="actions">
          <input type="submit"
                 value="Add Movie"
          >
          <input type="reset"
                 value="Clear Form"
          >
        </fieldset>
      </form>
    </main>
  </body>
</html>
`;

exports[`update a movie when: a user wants to view the form to edit a movie then: we return the view of the form 1`] = `
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <title>
    </title>
    <meta charset="UTF-8">
    <link rel="stylesheet"
          href="/stylesheets/application.css"
    >
  </head>
  <body>
    <header>
      <h1>
        <a href="/">
          The Ultimate Movie App
        </a>
      </h1>
    </header>
    <main>
      <form action="/movies/update/21"
            method="post"
      >
        <dl>
          <dt>
            <label for="title">
              Edit the title of the movie:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="title"
                   name="title"
                   placeholder="Title"
            >
          </dd>
          <dt>
            <label for="year">
              Edit the movie's year of release:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="year"
                   name="year"
                   placeholder="Release Year"
            >
          </dd>
          <dt>
            <label for="runTime">
              Edit the movie's run time:
            </label>
          </dt>
          <dd>
            <input type="text"
                   id="runTime"
                   name="runTime"
                   placeholder="Run time"
            >
          </dd>
        </dl>
        <fieldset class="actions">
          <input type="submit"
                 value="Update Movie"
          >
          <input type="reset"
                 value="Clear Form"
          >
        </fieldset>
      </form>
    </main>
  </body>
</html>
`;
